### ISSUE TITLE
Beyond Model Families: Summarizing Across Multiple Models

### ISSUE THREAD

**Issue by tmastny** [alexpghayes/modelling-in-r#32]
I would love to hear your thoughts on summaries across models. I think there is a lot of potential in this area, and would like to start thinking about sensible conventions. Full disclosure, I did a proof of concept on this months ago: [leadr](https://github.com/tmastny/leadr).

## Model and Model Family Summaries

In R, I think the backbone of model summaries exists in the `print` and `summary` methods you discussed [here](https://alexpghayes.github.io/modelling-in-r/using-models-programmatic-automated-mode.html), especially for interactive development. 

Scikit-learn is weak on summarized reporting. For example, something like this (which is taken from the [docs]()) shouldn't be best practice:

```python
print("Average difference of {0:6f} with std. dev. of {1:6f}."
      .format(score_difference.mean(), score_difference.std()))
```
On the other hand, `caret` does a lot better by providing a nice model summary right after training. 

```
## Linear Regression 
## 
## 381 samples
##  13 predictor
## 
## No pre-processing
## Resampling: Bootstrapped (25 reps) 
## Summary of sample sizes: 381, 381, 381, 381, 381, 381, ... 
## Resampling results:
## 
##   RMSE      Rsquared   MAE     
##   4.284876  0.7778135  2.955764
## 
## Tuning parameter 'intercept' was held constant at a value of TRUE
```

I think creating conventions for `print` and `summary` methods (also making it easy for `broom::tidy`) is an important consideration for modeling packages.

However, in general these methods only operate on a per model, or per model family basis. I think there is a lot of potential to standardize summary methods across models to help us efficiently compare or [ensemble](https://alexpghayes.github.io/modelling-in-r/interface-desirables.html#desirables-necessary-for-practicality) them. 

## Meta-Parameter Model Summaries

You made a great distinction between models and model families, but I think we can take this even further with *data models*.

In my mind data models (there probably exists a better term) operate in a new space of meta-parameters. These meta-parameters include the model type (random forest, glm, etc.), but also resampling method for tuning (and accuracy), resampling number, tuning metric, etc. 

If we can define these meta-parameters and make them consistent and accessible across models, we can aggregate these parameters. For example, `caret` has a consistent interface for these (but it isn't super accessible).

Then we can group these meta-parameters. Here's the result of [leadr](https://tmastny.github.io/leadr/articles/introduction.html) grabbing those meta-parameters off `caret` models:
```r
#> # A tibble: 5 x 13
#>    rank    id dir     model  metric score public method   num group index 
#>   <dbl>  <id> <chr>   <chr>  <chr>  <dbl>  <dbl> <chr>  <dbl> <dbl> <list>
#> 1    1.     2 models… rf     Accur… 0.955     NA boot     25.    2. <list…
#> 2    2.     5 models… glmnet Accur… 0.952     NA boot     25.    1. <list…
#> 3    3.     4 models… rf     Accur… 0.950     NA boot     25.    4. <list…
#> 4    4.     3 models… rf     Accur… 0.946     NA boot     25.    3. <list…
#> 5    5.     1 models… rf     Accur… 0.942     NA boot     25.    1. <list…
#> # ... with 2 more variables: tune <list>, seeds <list>
```

The second benefit of being able to group data models is selection for ensembles. You can choose models based on the meta-parameters and also have models grouped in a convenient list to send to the ensembling function. I've written about this in the past [here](https://tmastny.github.io/leadr/articles/ensemble.html).

**Comment by alexpghayes** [comment]
Sorry for the delay in responding. Partially this is due to busyness, partially I don't have much that I can add at the moment. I no longer think that the outline I've made in this repo of models / model families is the full picture, and currently am in the process of thinking through that some more.

If you take a look at https://tiny-url.com/broom-talk, I've started thinking about models (as I call them in this repo, or fits as I call them in the slides) as being akin to databases with tables for parameter, goodness of fit measures, and observation level information. `leadr` might then fit into a larger grammar as a databases of models (which themselves could be databases / tables). On its own I don't think this is especially interesting -- basically we say that there's a collection of models and all should implement a method that gives back some performance metric.

Perhaps most interesting is to consider what methods you might want to define on a **collection of models**. For prediction problems, one obvious answer seems to be various ensembling methods. At the moment I'm more curious about inference, however. Here are some questions to inspire thought on dealing with collections of models:

- what kinds of inference can you make from a collection of models? van der Laan's group has been working on TMLE which is one answer, but reading their stuff is rough and I don't really follow that well
- should all the models in the database be required to be fit on the same data? what about different representations of the original data? or subsets of the original data?

Most model selection tools (here I mean for inference, where the measure attempts to provide some guarantees about recovering a true model as the data goes to infinity) are for nested models. This is pretty boring to me: not only do you have to compare within model families, often you can only compare models trained using the same estimator.

Another thing to think about is model misspecification. Maybe you define some database/set of models and then you test them for misspecification and throw a bunch of them out. If you can define some automated misspecification tests, then I think this at least is an interesting method to a collection of models. Actually, I'm not strictly sure if this is method on a set of models, or applying a method on a single model to the entire set. In some sense this depends on whether information from model A can be used to check if model B is misspecified. I would argue that this is the case, but don't think there's much that could be done in terms of writing code for stuff. In another sense, the set of not-rejected-models is a fundamentally different from just being a set of models in that they most plausibly contain/approximate the truth.

There are also examples where you can do inference using many models at once. For example, you might ask which feature enters the LASSO path first. But again, each model has to be from the same model, and be fit using the same estimator.

This is a bit of brain barf at the moment, but the short version is that I don't have strong opinions about what operations make sense on a collection of models. Pretty much everything we do at the moment is to map some operation across the set, rather than operating on all elements of the set in a synthetic way.

**Comment by tmastny** [comment]
Database of models is exactly the type of phrase I was looking for. With regards to inference and collections of models, I am also not aware of any real significance.

However, I do think it can help solve an *organizational* issue, which I believe can be significantly improved in the interactive modeling workflow. Having models exist in the global namespace is a not a good way to understand what you've done, at least in my experience.

At the basic level, I think providing a modeling log (which `leadr` essentially does) can help gather and document your work for further analysis.

**Comment by alexpghayes** [comment]
Yeah I can't wait to use `leadr` for my next predictive modeling task! It definitely automates some useful steps.

I know that some other people more on the CS side of things are thinking about the logging side of things, but I don't know how much of that is for debugging vs. usability.
